src/Quasar/Advanced/Types.purs

```diff
@@ -0,0 +1,393 @@
+        <|>
+        (map Dir $ lmap (const $ "Incorrect directory resource") $ parseDir pt)
+parseFile ∷ String → Either String (Pt.AbsFile Pt.Sandboxed)
+parseFile pt =
+  Pt.parseAbsFile pt
+  >>= Pt.sandbox Pt.rootDir
+  <#> (Pt.rootDir </> _)
+  # maybe (Left "Incorrect resource") pure
+
+parseDir ∷ String → Either String (Pt.AbsDir Pt.Sandboxed)
+parseDir pt =
+  Pt.parseAbsDir pt
+  >>= Pt.sandbox Pt.rootDir
+  <#> (Pt.rootDir </> _)
+  # maybe (Left "Incorrect resource") pure
+
+newtype Action =
+  Action
+    { operation ∷ Operation
+    , resource ∷ Resource
+    , resourceType ∷ ResourceType
+    }
+
+instance encodeJsonAction ∷ EncodeJson Action where
+  encodeJson (Action obj) =
+    "operation" := obj.operation
+    ~> "resource" := obj.resource
+    ~> "resourceType" := obj.resourceType
+    ~> jsonEmptyObject
+
+instance decodeJsonAction ∷ DecodeJson Action where
+  decodeJson = decodeJson >=> \obj →
+    { operation: _
+    , resource: _
+    , resourceType: _
+    }
+    <$> (obj .? "operation")
+    <*> (obj .? "resource")
+    <*> (obj .? "resourceType")
+    <#> Action
+
+
+newtype UserId = UserId String
+runUserId ∷ UserId → String
+runUserId (UserId s) = s
+
+instance encodeJsonUserId ∷ EncodeJson UserId where
+  encodeJson = runUserId >>> encodeJson
+
+instance decodeJsonUserId ∷ DecodeJson UserId where
+  decodeJson = map UserId <<< decodeJson
+
+
+newtype TokenId = TokenId String
+runTokenId ∷ TokenId → String
+runTokenId (TokenId s) = s
+
+instance encodeJsonTokenId ∷ EncodeJson TokenId where
+  encodeJson = runTokenId >>> encodeJson
+
+instance decodeJsonTokenId ∷ DecodeJson TokenId where
+  decodeJson = map TokenId <<< decodeJson
+
+newtype PermissionId = PermissionId String
+runPermissionId ∷ PermissionId → String
+runPermissionId (PermissionId s) = s
+
+instance encodeJsonPermissionId ∷ EncodeJson PermissionId where
+  encodeJson = runPermissionId >>> encodeJson
+
+instance decodeJsonPermissionId ∷ DecodeJson PermissionId where
+  decodeJson = map PermissionId <<< decodeJson
+
+
+data GrantedTo
+  = UserGranted UserId
+  | GroupGranted (Pt.AbsFile Pt.Sandboxed)
+  | TokenGranted TokenId
+
+instance encodeJsonGrantedTo ∷ EncodeJson GrantedTo where
+  encodeJson (UserGranted uid) = encodeJson uid
+  encodeJson (GroupGranted pt) = encodeJson $ Pt.printPath pt
+  encodeJson (TokenGranted tk) = encodeJson tk
+
+instance decodeJsonGrantedTo ∷ DecodeJson GrantedTo where
+  decodeJson = decodeJson >=> \str →
+    (map GroupGranted $ parseFile str)
+    <|>
+    (map (UserGranted <<< UserId) $ checkUserId str)
+    <|>
+    (pure $ TokenGranted $ TokenId str)
+    where
+    checkUserId ∷ String → Either String String
+    checkUserId str =
+      -- ¯\_(ツ)_/¯
+      if isJust (Str.indexOf "@" str)
+        then pure str
+        else Left "Incorrect email"
+
+newtype GrantedBy =
+  GrantedBy
+    { tokens ∷ Array TokenId
+    , users ∷ Array UserId
+    , groups ∷ Array (Pt.AbsFile Pt.Sandboxed)
+    }
```

Comment:
I'd prefer to avoid `newtype`s around records in favour of plain records if we can, to match the other cases where this occurs.

Unless these records are internal to the quasar stuff and not actually useful for consumers of the API, in which case `newtype` away if it makes things easier!